src/aspire/source/__init__.py

```diff
@@ -370,6 +372,108 @@ def whiten(self, noise_filter):
+
+        logger.info('Apply contrast reverse on source object')
+        L = self.L
+        image_center = (L + 1) / 2
+        coor_mat_m, coor_mat_n = np.meshgrid(np.arange(1, L + 1), np.arange(1, L + 1))
+
+        distance_from_center = np.sqrt((coor_mat_m - image_center) ** 2
+                                       + (coor_mat_n - image_center) ** 2)
+
+        # Get mask indices of signal and noise samples assuming molecule is around the center
+        signal_mask = distance_from_center < round(L / 4)
+        noise_mask = distance_from_center > round(L / 2 * 0.8)
+
+        # Calculate mean values in batch_size
+        signal_mean = 0.0
+        noise_mean = 0.0
+        for i in range(0, self.n, batch_size):
+            images = self.images(i, batch_size).asnumpy()
+            signal = (images * np.expand_dims(signal_mask, 2))
+            noise = (images * np.expand_dims(noise_mask, 2))
+            signal_denominator = self.n * np.sum(signal_mask)
+            noise_denominator = self.n * np.sum(noise_mask)
+            signal_mean += np.sum(signal) / signal_denominator
+            noise_mean += np.sum(noise) / noise_denominator
+
+        if signal_mean < noise_mean:
+            logger.info('Need to reverse contrast')
+            scale_factor = -1.0 * np.ones(self.n)
+        else:
+            logger.info('No need to reverse contrast')
+            scale_factor = 1.0 * np.ones(self.n)
+
+        logger.info('Adding Scaling Xform to end of generation pipeline')
+        self.generation_pipeline.add_xform(Multiply(scale_factor))
+
+        # Invalidate images
+        self._im = None
+
+    def normalize_background(self, radius=None, batch_size=512):
+        """
+        Normalize the images by the noise background
+
+        This is done by shifting the image density by the mean value of background
+        and scaling the image density by the standard deviation of background.
+        From the implementation level, we modify the `ImageSource` in-place by
+        appending the `Reduce` and `Multiple` filters to the generation pipeline.
+
+        :param batch_size: Batch size of images to query.
+        :return: On return, the `ImageSource` object has been modified in place.
+        """
+        logger.info('Normalize background on source object')
+        L = self.L
+        radius = L // 2 if radius is None else radius
+
+        grid = grid_2d(L)
+        mask = grid['r'] > (radius / L)
+
+        mean = 0.0
+        variance = 0.0
+        for i in range(0, self.n, batch_size):
+            images = self.images(start=i, num=batch_size).asnumpy()
+            images_masked = (images * np.expand_dims(mask, 2))
+            mean += np.sum(images_masked)
+            variance += np.sum(np.abs(images_masked**2))
+        denominator = self.n * np.sum(mask)
+        mean /= denominator
+        variance /= denominator
+        std = np.sqrt(variance)
+        logger.info('Adding Reduce Xform to end of generation pipeline')
```

Comment:
Space these statements out a bit. It's a little tight right now.

Reply:
Done.